coda-src/venus/fso_cachefile.cc

```diff
@@ -575,35 +575,54 @@ CacheChunkList::~CacheChunkList()
 
 CacheChunk CacheFile::GetNextHole(uint64_t start_b, uint64_t end_b)
 {
-    /* Floor length in blocks */
-    uint64_t length_b_f = bytes_to_ccblocks_floor(length);
-    /* Ceil length in blocks */
-    uint64_t length_b  = bytes_to_ccblocks_ceil(length);
-    uint64_t holestart = start_b;
-    int64_t holesize   = 0;
+    CODA_ASSERT(start_b <= end_b);
+    /* Number of blocks of the cache file */
+    uint64_t nblocks = bytes_to_ccblocks_ceil(length);
+    /* Number of full blocks */
+    uint64_t nblocks_full   = bytes_to_ccblocks_floor(length);
+    uint64_t hole_start_idx = 0;
+    uint64_t hole_end_idx   = 0;
+    uint64_t hole_size      = 0;
+
+    CODA_ASSERT(start_b <= nblocks);
+
+    if (end_b > nblocks) {
+        end_b = nblocks;
+    }
 
-    for (uint64_t i = start_b; i < end_b; i++) {
-        if (cached_chunks->Value(i)) {
-            holesize  = 0;
-            holestart = i + 1;
-            continue;
+    /* Find the start of the hole */
+    for (hole_start_idx = start_b; hole_start_idx < end_b; hole_start_idx++) {
+        if (!cached_chunks->Value(hole_start_idx)) {
+            break;
         }
+    }
+
+    /* No hole */
+    if (hole_start_idx == end_b)
+        return CacheChunk();
 
-        /* The last block might not be full */
-        if (i + 1 == length_b) {
-            holesize += length - (length_b_f << CacheChunkBlockSizeBits);
-            return (CacheChunk(holestart * CacheChunkBlockSize, holesize));
+    /* Find the end of the hole */
+    for (hole_end_idx = hole_start_idx; hole_end_idx < end_b; hole_end_idx++) {
+        if (cached_chunks->Value(hole_end_idx)) {
+            break;
         }
+    }
 
-        /* Add a full block */
-        holesize += CacheChunkBlockSize;
+    CODA_ASSERT(hole_end_idx > hole_start_idx);
 
-        if ((i + 1 == end_b) || cached_chunks->Value(i + 1)) {
-            return (CacheChunk(holestart * CacheChunkBlockSize, holesize));
-        }
+    hole_size = ccblocks_to_bytes(hole_end_idx - hole_start_idx - 1);
+
+    /* If the hole ends at EOF and it has a tail */
+    if (hole_end_idx == nblocks && nblocks_full != nblocks) {
+        /* Add the tail */
+        uint64_t tail = length - ccblocks_to_bytes(nblocks - 1);
```

Comment:
should this be ccblocks_to_bytes(nblocks_full -1) ?

Reply:
actually, nblocks-1 should be ok, at this point I believe the following holds true
 nblocks-1 == nblocks_full
which makes ccblocks_to_bytes(nblocks-1) the length rounded down to the start of the chunk